web-api/src/storage/elastic.rs

```diff
@@ -396,22 +396,51 @@ impl storage::Document for Storage {
         params: KnnSearchParams<'a>,
     ) -> Result<Vec<models::PersonalizedDocument>, Error> {
         // https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn
+        let excluded_ids = json!({
+            "values": params.excluded.iter().map(AsRef::as_ref).collect_vec()
+        });
+
+        let filter = if let Some(published_after) = params.published_after {
+            // published_after != null && published_after <= publication_date <= now
+            json!({
+                "bool": {
+                    "filter": {
+                        "range": {
+                            "properties.publication_date": {
+                                "gte": published_after.to_rfc3339(),
+                                "lte": "now"
+                            }
+                        }
+                    },
+                    "must_not": {
+                        "ids": excluded_ids
+                    }
+                }
+            })
+        } else {
+            // published_after == null || published_after <= now
+            json!({
+                "bool": {
+                    "must_not": [
+                        { "ids": excluded_ids },
+                        { "range": {
+                            "properties.publication_date": {
+                                "gt": "now"
+                            }
+                        }}
+                    ]
+                }
+            })
+        };
+
         let body = Some(json!({
             "size": params.k_neighbors,
             "knn": {
                 "field": "embedding",
                 "query_vector": params.embedding,
                 "k": params.k_neighbors,
                 "num_candidates": params.num_candidates,
-                "filter": {
-                    "bool": {
-                        "must_not": {
-                            "ids": {
-                                "values": params.excluded.iter().map(AsRef::as_ref).collect_vec()
-                            }
-                        }
-                    }
-                }
+                "filter": filter,
```

Comment:
:art: no trailing commas in json
```suggestion
                "filter": filter
```